File: frontend/src/objects/problem/ProblemStatementText.ts
/** 题目题面正文品牌类型；用于 markdown/文本题面内容。 */
export type ProblemStatementText = string & { readonly __brand: 'ProblemStatementText' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建题面正文品牌值；调用前必须完成非空和长度校验。 */
function createProblemStatementText(value: string): ProblemStatementText {
  /** 注意：这里的 as 只在 parseProblemStatementText 校验通过后施加品牌类型。 */
  return value as ProblemStatementText
}

/** 将题面正文品牌值还原为字符串；用于 API body 和编辑器初始值。 */
export function problemStatementTextValue(statement: ProblemStatementText): string {
  return statement
}

/** 解析题面正文；去除首尾空白并限制最大长度。 */
export function parseProblemStatementText(rawStatement: string): ParseResult<ProblemStatementText> {
  const normalized = rawStatement.trim()
  if (!normalized) {
    return { ok: false, error: 'Problem statement is required.' }
  }
  if (normalized.length > 20000) {
    return { ok: false, error: 'Problem statement must be at most 20000 characters.' }
  }
  return { ok: true, value: createProblemStatementText(normalized) }
}

File: backend/src/main/scala/domains/problem/objects/ProblemStatementText.scala
package domains.problem.objects

import io.circe.{Decoder, Encoder}


/** 题面正文文本；后端只校验非空和长度，不解释具体 Markdown/HTML 语义。 */
final case class ProblemStatementText(value: String)

/** 题面正文的 JSON 编解码与输入校验入口。 */
object ProblemStatementText:
  given Encoder[ProblemStatementText] = Encoder.encodeString.contramap(_.value)
  given Decoder[ProblemStatementText] = Decoder.decodeString.emap(parse)

  /** 规范化并校验题面正文；空文本和超长文本会被拒绝。 */
  def parse(raw: String): Either[String, ProblemStatementText] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Problem statement is required.")
    else if normalized.length > 20000 then Left("Problem statement must be at most 20000 characters.")
    else Right(ProblemStatementText(normalized))
